### PR TITLE
Add Python 3.12 compatibility for path fieldtype

### LINKED ISSUE #89: `_PosixFlavour` is removed in Python 3.12
Python 3.12 removed `pathlib._PosixFlavour` which we use in a unit test and now fails. We also use the `_flavour` attribute but that still seems to exist.

### PR DESCRIPTION
**Pull request by yunzheng**
Fixes #89

### CODE REVIEW

File: tests/test_fieldtypes.py [hunk 1]
```diff
@@ -526,10 +527,18 @@ def test_digest():
         excinfo.match(r".*Invalid MD5.*")
 
 
+if PY_312:
+    # Starting from Python 3.12, pathlib._Flavours are removed as you can now properly subclass pathlib.Path
+    # See also: https://github.com/python/cpython/issues/88302
+    PosixFlavour = pathlib.PurePosixPath
```
Comment (pyrco) [inline tests/test_fieldtypes.py:533]: The `_flavour` property in 3.12 is a link to the `posixpath` or `ntpath` module. We could do something similar by having a function `def custom_flavour(sep, altsep)` that yields a `patch()`ed posixpath with the given separator and alt separator and modify `PureCustomPath` like:
```
    class PureCustomPath(pathlib.PurePath):
        if PY_312:
            _flavour = custom_flavour(sep, altsep)
        else:
            _flavour = CustomFlavour()
```

Our tests however just need a PurePath subclass instance with a `_flavour` property that has `sep` and `altsep` properties, so a simpler setup would for this case also work.

**Review by pyrco — CHANGES_REQUESTED** [review]

File: flow/record/fieldtypes/__init__.py [hunk 2]
```diff
@@ -675,7 +676,11 @@ def __new__(cls, *args):
                     continue
                 break
 
-        return cls._from_parts(args)
+        if PY_312:
+            obj = super().__new__(cls)
```
[the remark this answers is not in the document]
Comment (yunzheng) [reply flow/record/fieldtypes/__init__.py:680]: i'm happy to accept any suggestions

File: tests/test_fieldtypes.py (hunk 1 above)
Comment (yunzheng) [reply tests/test_fieldtypes.py:533]: i'm happy to accept any suggestions

Comment (pyrco) [reply tests/test_fieldtypes.py:533]: I'd say, just do
```suggestion
    PosixFlavour = Mock
```

as we probably don't want to really mock `posixpath` and now it is still clear that flavour is something different.

Comment (yunzheng) [reply tests/test_fieldtypes.py:533]: This change gives me an error:

```python
.tox/py3/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:178: in exec_module
    exec(co, module.__dict__)
tests/test_fieldtypes.py:558: in <module>
    (custom_pure_path(sep="/", altsep="")("/foo/bar"), True),
tests/test_fieldtypes.py:547: in custom_pure_path
    class PureCustomPath(pathlib.PurePath):
tests/test_fieldtypes.py:548: in PureCustomPath
    _flavour = CustomFlavour()
tests/test_fieldtypes.py:543: in __new__
    instance.sep = sep
/usr/local/lib/python3.12/unittest/mock.py:771: in __setattr__
    elif (self._spec_set and self._mock_methods is not None and
/usr/local/lib/python3.12/unittest/mock.py:656: in __getattr__
    elif self._mock_methods is not None:
/usr/local/lib/python3.12/unittest/mock.py:655: in __getattr__
    raise AttributeError(name)
E   AttributeError: _mock_methods
```

File: flow/record/fieldtypes/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yunzheng) [reply flow/record/fieldtypes/__init__.py:682]: Added suggestion in new commit

File: tests/test_fieldtypes.py (hunk 1 above)
Comment (yunzheng) [reply tests/test_fieldtypes.py:533]: @pyrco any ideas?

**Comment by pyrco** [comment]
I had this change in mind. It's not a real mock of `posixpath`/`ntpath` but for the testcase purposes it is sufficient.

**Comment by pyrco** [comment]
Ok, this didn't work as expected, I'll have to look a bit more into it

**Comment by yunzheng** [comment]
> Ok, this didn't work as expected, I'll have to look a bit more into it

any luck with this?

**Comment by pyrco** [comment]
Yes, that is the mock I was looking for :wink:

**Review by pyrco — APPROVED** [review]

**Pull request merged**